### ISSUE TITLE
It's impossible to disconnect from socket if it's in reconnect state (phoenix.js)

### ISSUE THREAD

**Issue by vheathen** [phoenixframework/phoenix#1841]
Hi!

  I'm trying to use phoenix channels as a transport and found weird behaviour.

  If connect to the socket with phoenix.js and then kill the server it is not possible to "disconnect" (prevent additional reconnect tries) until connection reestablished.

### Environment

* NodeJS version (node -v): v5.11.0
* NPM version (npm -v): 3.8.6
* Operating system: Fedora 23 + Chrome 52.0.2743.82 (64-bit)

Code snippet:
```(javascript)
import { Component } from '@angular/core';
import { Config } from '../../config';
import { Socket } from 'phoenix/priv/static/phoenix.js';

@Component({
             selector:      'sg-test',
             template:      `
<form>
  <div class="form-group">
    <button type="button" (click)="connect()" class="btn btn-default">Connect</button>
    <button type="button" (click)="disconnect()" class="btn btn-default">Disconnect</button>
  </div>
</form>    `
           })

export default class TestComponent {

  private socket: any;

  private connect() {
    this.socket = new Socket(Config.SOCKET_USER_URL, {
      params: {token: localStorage.getItem(Config.JWT_TOKEN_NAME)}
    });
    this.socket.connect();
    console.log('Connect ', this.socket);
  }

  private disconnect() {
    this.socket.disconnect();
    console.log('Disconnect ', this.socket);
  }
}

```

### Expected behavior
Should stop reconnecting after `disconnect()` call

### Actual behavior
Continue to reconnect even after `disconnect()` call - it can be confirmed in browser console.

I think disconnect() method should interrupt reconnect timer but I'm not much familiar with JS to publish a PR.

I also think that socket disconnect and channel leave should unsubscribe from events as well but probably it's another story.

Thank you and best regards,
Vladimir

**Comment by chrismccord** [comment]
You did not specify the Phoenix version. Can you confirm you are on the latest phoenix version (and phoenix.js client if you are venturing it or grabbing from npm/elsewhere) ?

**Comment by vheathen** [comment]
Oh, sorry. Yes, it's the phoenix package v1.2.0 from npm.

**Comment by vheathen** [comment]
@chrismccord thanks for the reply and sorry if I wasn't clear enough. To be honest, I can't reproduce your experiment: in my case I can connect to the WS socket with a raw WS (can see that it's connected to the server with debug logging), shutdown the server node (at this time socket is closing by itself with `onclose` callback) and call socket.close() without any additionals errors. Probably I do something wrong:
Open a new Chrome window and set URL to about:blank. Open console.
```
> var socket = new WebSocket("ws://localhost:4000/device/websocket")
undefined
WebSocket connection to 'ws://localhost:4000/device/websocket' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED
```
The error message appears only once.

Start the server node
```
> var socket = new WebSocket("ws://localhost:4000/device/websocket")
undefined
> socket.onerror = (error) => { console.log('Error: ', error) }
function (error) => { console.log('Error: ', error) }
> socket.onclose = (event) => { console.log('Closed: ', event) }
function (event) => { console.log('Closed: ', event) }
```
interrupt the server node
```
VM6770:1 Closed:  CloseEvent {wasClean: false, code: 1006, reason: "", type: "close", target: WebSocket…}
```
The last message appears only once - the socket doesn't reconnect by itself.

In case of `socket.close()` **before** the node shutdown I've got
```
> socket.close()
undefined
VM89:1 Closed:  CloseEvent {wasClean: true, code: 1005, reason: "", type: "close", target: WebSocket…}
```
and also no reconnect tries\errors even in case of the following server shutdown.

But in my original message I've meant the next sequence: connect to the Phoenix socket and channel, shutdown the server before Phoenix socket disconnect() function call, and after that it's impossible to prevent additional reconnect tries from the Phoenix socket - it will try to reconnect forever (until the next success) even if one intentionally calls disconnect(). I believe it's because there is no option to reset the reconnectTimer in the Socket.disconnect() function.

Btw, in case of the node failure or shutdown while phoenix socket is connected there is the following message in logs (if one puts 'logger' option to the socket constructor):
```
transport: close CloseEvent {wasClean: false, code: 1006, reason: "", type: "close", target: WebSocket…}
```

**Comment by merqlove** [comment]
Looks like it's enough to exit with #close(1000).
I think connection is closed, but it's  browser decision, to-drop/not-to-drop old socket from the list. 
I think it's ok if the old socket still persists in the list. It's give us some history for debug.
